### DISCUSSION TITLE
How to change user who committed on the GitHub repo

### DISCUSSION THREAD

**Discussion by badal8381** [community/community#61419] [Repositories] [answered]
I was working on multiple projects. I committed a repo with different user by mistake and it is in past 3-4 days ago. Is there any way I can change just the user.

A quick reply from anyone is appreciated

**Comment by nirav-gajera** [comment, marked as answer]
Yes, you can change the user for a commit in GitHub. Here are two methods:

**Method 1: Using Interactive Rebase**

1. Navigate to the directory of the repo that you want to change the user for.
2. Run the following command:

```
git rebase -i HEAD~3
```

This will open your editor with a list of all the commits that have been made in the past 3 days.
3. For each commit that you want to change the user for, add the `edit` keyword to the line.
4. Save and close your editor.
5. Git will interactively rebase your commits, giving you the opportunity to change the user for each commit.
6. Once you have finished changing the users, run the following command to push the changes to GitHub:

```
git push
```

**Method 2: Using the `git filter-branch` Command**

1. Navigate to the directory of the repo that you want to change the user for.
2. Run the following command:

```
git filter-branch --env-filter 'OLD_EMAIL=old_email NEW_NAME=new_name NEW_EMAIL=new_email' HEAD~3..
```

This will replace all occurrences of the old email address and username with the new email address and username in all commits that have been made in the past 3 days.
3. Once the command has finished, run the following command to push the changes to GitHub:

```
git push
```

Which method you use will depend on your specific needs. If you only need to change the user for a few commits, then Method 1 is a good option. If you need to change the user for a large number of commits, then Method 2 is a better option.

Here are some additional resources that you may find helpful:

* How to Change the Author (Name / Email) of a Commit: https://www.git-tower.com/learn/git/faq/change-author-name-email
* Changing Your Committer Name & Email per Repository: https://www.git-tower.com/learn/git/faq/change-author-name-email
* Why are my commits linked to the wrong user?: https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user

I hope this helps! Let me know if you have any other questions.

**Comment by badal8381** [comment]
thank you for the response. will this method change the commit hash or history

**Reply by nirav-gajera** [reply]
Yes, changing the commit message, file changes, commit author, date, or parent commit hash will change the commit hash. This is because the commit hash is a unique identifier that is generated based on all of these factors. If any of these factors change, the commit hash will also change.

The commit history is a record of all of the commits that have been made to a repository. It is a chronological list of the commit hashes, along with the commit messages and other metadata. If you change the commit hash, the commit history will also be changed to reflect the new hash.

**Comment by badal8381** [comment]
thank you so much for help
